rfcs/0003-storage.md

```diff
@@ -1,46 +1,108 @@
+
+## 6.3 Validator behavior
+
+1. Validator joins the network and submits a storage mining capacity transaction which tells the network how many proofs it can process in a given period defined by NUM\_KEY\_ROTATION\_TICKS.
+2. Every NUM\_KEY\_ROTATION\_TICKS the validator stores the PoH value at that height.
+4. Every NUM\_KEY\_ROTATION\_TICKS it also validates samples received from replicators. It signs the PoH hash at that point and uses the low 5 bits of the first byte to index into the signature and then picks a number of bytes of the received sha states with that one in choosing which storage proofs to validate. It then creates a mask of valid proofs and fake proofs and sends it to the leader. This is a storage proof confirmation transaction.
+5. The storage proof confirmation transaction is integrated into the ledger.
+6. After a lockout period of NUM\_SECONDS\_STORAGE\_LOCKOUT seconds, the validator then submits a storage proof claim transaction which then causes the distribution
+of the storage reward if no challenges were seen for the proof to the validators and replicators party to the proofs.
+7. Validator responds to RPC interfaces for what its encryption key state is and PoH values for any rotation events.
+
+## 6.4 Replicator behavior
+
+1. Replicator paranoia modes:
+For many of the operations there are number of options depending on how paranoid a replicator is:
+- (a) replicator can ask a validator
+- (b) replicator can ask multiple validators
+- (c) replicator can subscribe to the full transaction stream and generate the key state itself
+- (d) replicator can subscribe to an abbreviated transaction stream to generate the key state itself
+2. A replicator obtains the PoH hash corresponding to the last key rotation.
+3. The replicator signs the PoH hash with its keypair. That signature is the seed used to pick the block to replicate and also the encryption key.
+4. The replicator sends a request to a validator with its signature asking for which block it should now store. The validator will respond with the starting entry height and the number of entries.
+5. The replicator retrives the ledger by asking peer validators and replicators. See 6.4.
+6. The replicator then encrypts that block with the key with chacha algorithm in CBC mode with NUM\_CHACHA\_ROUNDS of encryption.
+7. The replicator asks the validator for the PoH hash corresponding its value for offset\_last\_id
+8. The replicator then signs that Poh hash with its keypair and initializes a chacha rng with the signature as the seed.
+9. The replicator generates NUM\_STORAGE\_SAMPLES samples in the range of the entry size and samples the encrypted block with sha256 for 32-bytes at each offset value. Sampling the state should be faster than generating the encrypted block.
+10. The replicator sends a PoRep proof transaction which contains its sha state at the end of the sampling operation, its seed and the samples it used to the current leader and it is put onto the ledger.
```

Comment:
how do we trust that the samples were actually sampled at the rate claimed in the proof transaction, if it is only written into PoH as a batch?

Reply:
It takes some amount of time to generate the samples and samples are continuously added to the chain. I think we are relying on the motivation of better chance for a reward that replicators will be generating storage proofs as fast as they can such to increase their chance of the highest reward.